src/libprojectM/Audio/MilkdropFFT.hpp

```diff
@@ -0,0 +1,159 @@
+ * Also applies an equalizer pattern with an envelope curve to the resulting data to smooth out
+ * certain artifacts.
+ */
+class MilkdropFFT
+{
+public:
+    /**
+     * Initializes the Fast Fourier Transform.
+     * @param samplesIn Number of waveform samples which will be fed into the FFT
+     * @param samplesOut Number of frequency samples generated; MUST BE A POWER OF 2.
+     * @param equalize true to roughly equalize the magnitude of the basses and trebles,
+     *                 false to leave them untouched.
+     * @param envelopePower Specifies the envelope power. Set to any negative value to disable the envelope.
+     *                      See InitEnvelopeTable for more info.
+     */
+    MilkdropFFT(size_t samplesIn, size_t samplesOut, bool equalize = true, float envelopePower = 1.0f);
+
+    /**
+     * @breif Converts time-domain samples into frequency-domain samples.
+     * The array lengths are the two parameters to Init().
+     *
+     * The last sample of the output data will represent the frequency
+     * that is 1/4th of the input sampling rate.  For example,
+     * if the input wave data is sampled at 44,100 Hz, then the last
+     * sample of the spectral data output will represent the frequency
+     * 11,025 Hz.  The first sample will be 0 Hz; the frequencies of
+     * the rest of the samples vary linearly in between.
+     *
+     * Note that since human hearing is limited to the range 200 - 20,000
+     * Hz.  200 is a low bass hum; 20,000 is an ear-piercing high shriek.
+     *
+     * Each time the frequency doubles, that sounds like going up an octave.
+     * That means that the difference between 200 and 300 Hz is FAR more
+     * than the difference between 5000 and 5100, for example!
+     *
+     * So, when trying to analyze bass, you'll want to look at (probably)
+     * the 200-800 Hz range; whereas for treble, you'll want the 1,400 -
+     * 11,025 Hz range.
+     *
+     * If you want to get 3 bands, try it this way:
+     *   a) 11,025 / 200 = 55.125
+     *   b) to get the number of octaves between 200 and 11,025 Hz, solve for n:
+     *          2^n = 55.125
+     *          n = log 55.125 / log 2
+     *          n = 5.785
+     *   c) so each band should represent 5.785/3 = 1.928 octaves; the ranges are:
+     *          1) 200 - 200*2^1.928                    or  200  - 761   Hz
+     *          2) 200*2^1.928 - 200*2^(1.928*2)        or  761  - 2897  Hz
+     *          3) 200*2^(1.928*2) - 200*2^(1.928*3)    or  2897 - 11025 Hz
+     *
+     * A simple sine-wave-based envelope is convolved with the waveform
+     * data before doing the FFT, to emeliorate the bad frequency response
+     * of a square (i.e. nonexistent) filter.
+     *
+     * You might want to slightly damp (blur) the input if your signal isn't
+     * of a very high quality, to reduce high-frequency noise that would
+     * otherwise show up in the output.
```

Comment:
is there any audio we expect people to feed in that will not be of "a very high quality"?

Reply:
Don't know, probably not relevant anymore today. Most of the Doxygen text in this class is just a 1:1 copy of the comments in Milkdrop's code, I didn't change it for "historical reasons". :grin: